### DISCUSSION TITLE
PX4/Ardupilot rate controller performance in GPS denied scenarios

### DISCUSSION THREAD

**Discussion by pallavbhalla** [ctu-mrs/mrs_uav_system#192] [Q&A]
Hi

Thanks for sharing this work.

This [**response**](https://github.com/ctu-mrs/mrs_uav_hw_api/issues/2#issuecomment-2462545279) brought a few questions regarding PX4/Ardupilot's body rate controller performance in GPS denied environments(like indoors, caves etc). 

Questions:

1. From the discussion and core estimator configs in mrs_uav_system, it seems like GPS is the primary source of position corrections in PX4 EKF. Am I right? Are there any secondary or tertiary sources configured in PX4 EKF?
2. AFAIK GPS data is used to estimate the IMU gyro and accel biases. Right? If GPS is denied/degraded(e.g. indoors) and LIO/VIO is the active estimator in EstimationManager, what happens to the body rate controller on PX4/Ardupilot which relies on IMU Gyro/Accel data? 
3.  How does the PX4 body rate controller cope with the IMU bias drifts? Won't it lead to degrading control performance if not handled?
4. **Won't we need external estimator(LIO/VIO etc.) corrections to be sent to PX4/Ardupilot apart from the desired body rates in order to correct the accumulated drifts? Do you foresee any challenges with this?**

Could you please clarify?

Thanks again

**Comment by petrlmat** [comment]
Hi @Pallav1299, I will try to answer your questions.

> From the discussion and core estimator configs in mrs_uav_system, it seems like GPS is the primary source of position corrections in PX4 EKF. Am I right? Are there any secondary or tertiary sources configured in PX4 EKF?

Yes that's correct. In fact, the configs of mrs_uav_system don't have any influence on the PX4 EKF. GPS is the primary source of the PX4 EKF and we aren't changing any parameters that would influence the position corrections in PX4 EKF, so if there are any other corrections, they are used as set by default. E.g. the PX4 EKF also uses a barometer for the position estimation by default.

> AFAIK GPS data is used to estimate the IMU gyro and accel biases. Right? If GPS is denied/degraded(e.g. indoors) and LIO/VIO is the active estimator in EstimationManager, what happens to the body rate controller on PX4/Ardupilot which relies on IMU Gyro/Accel data?

In this case, the body rate controller in PX4 uses the angular velocities from gyro data for feedback. The gyro bias might or might not be estimated using the specific force measurements and magnetometers when available. I'm not that familiar with the PX4, so I'm not sure whether there actually are any corrections of the angular rates in the body rate controller or not. Maybe someone more experienced with PX4 can clarify this?

> How does the PX4 body rate controller cope with the IMU bias drifts? Won't it lead to degrading control performance if not handled?

That's a great question, to which I don't know the exact answer. My guess is that the body rate controller is not aware of the gyro bias so it has a near zero control error. However, higher-level controllers (e.g. attitude) have to compensate for the error caused by the gyro bias, which they most likely do by employing some integral components in the control function.

> Won't we need external estimator(LIO/VIO etc.) corrections to be sent to PX4/Ardupilot apart from the desired body rates in order to correct the accumulated drifts? Do you foresee any challenges with this?

I personally don't have any experience with sending corrections to PX4, so I'm not sure what is necessary to achieve that. Please share your progress in this regard in this thread for all others who might be interested in this feature.

Our goal is to not couple our system with PX4/Ardupilot/Betaflight or any other FCUs or their firmware too much to keep the system general, modular, and able to work with the most FCUs available. So sending corrections back to the FCU or tinkering with the PX4 EKF would be a step in the opposite direction for us.
